WT-10898 Add assertions for owning a spinlock

pmacko86 opened the pull request.

agorrod commented:
I was browsing the open PRs and came across this - it's a nice improvement, thanks!

sueloverso commented on src/include/mutex_inline.h:
```diff
@@ -266,6 +292,40 @@ __wt_spin_unlock(WT_SESSION_IMPL *session, WT_SPINLOCK *t)
+ * WT_ASSERT_SPINLOCK_COND --
+ *      Assert that the session owns the spinlock, but only if the given condition is satisfied.
+ */
+#define WT_ASSERT_SPINLOCK_COND(session, t, condition) \
```
I would personally rather see the WT_ASSERT statement in the code rather than wrapped in this macro. It reads more explicitly to me, and feels clearer for the conditional one. It also allows for flexibility for future work that might want to use one of the other WT_ASSERT macros like WT_ASSERT_ALWAYS or WT_ASSERT_OPTIONAL without having to clutter up a lot more wrapper macros.

pmacko86 replied:
To clarify, is it okay to keep the WT_ASSERT_SPINLOCK macro, even as we remove this _COND version? I'd really like to keep WT_ASSERT_SPINLOCK as it makes the assert a lot less verbose.

If we remove the _COND version, would it be okay to have something like this?
```
if (!F_ISSET(S2BT(session), WT_BTREE_SALVAGE))
    WT_ASSERT_SPINLOCK(session, &block->live_lock);
```
This would be probably the most readable way to do this. And I'd like to believe that a good compiler would figure out that it can remove even the `if` check in the release mode, because it has no side effect.

I originally had this with just WT_ASSERT calling the _owns function directly, but I found the condition in it not particularly readable, so I made this one.

sueloverso commented on src/include/mutex_inline.h:
```diff
@@ -266,6 +292,40 @@ __wt_spin_unlock(WT_SESSION_IMPL *session, WT_SPINLOCK *t)
+ *     Check whether the session owns the spinlock.
+ */
+static inline bool
+__wt_spin_owns(WT_SESSION_IMPL *session, WT_SPINLOCK *t)
```
This is very minor but it seemed like `owned` goes better with `locked` in my head.

sueloverso commented on src/block/block_ckpt.c:
```diff
@@ -828,8 +827,7 @@ __ckpt_process(WT_SESSION_IMPL *session, WT_BLOCK *block, WT_CKPT *ckptbase)
 
-    if (locked)
-        __wt_spin_unlock(session, &block->live_lock);
+    __wt_spin_unlock_if_owns(session, &block->live_lock);
```
This is a nice cleanup.

quchenhao commented on src/include/session.h:
```diff
@@ -53,6 +53,12 @@ typedef TAILQ_HEAD(__wt_cursor_list, __wt_cursor) WT_CURSOR_LIST;
+#define WT_SESSION_ID_INVALID 0xffffffff
+
+/* A fake session ID for when we need to refer to a session that is actually NULL. */
+#define WT_SESSION_ID_NULL 0xfffffffe
```
I was thinking the session is actually null in this case but apparently I was wrong. Who set the session id to this value and why we don't have this definition before the pr?

pmacko86 replied:
I just added these two special session IDs that I ran into:

- given that the session ID is defined as an index into an array, 0 is thus considered a valid session ID
- we actually call lock/unlock in two places with session being NULL, for which I also needed to define an ID value for the purposes of the locks

I suppose we didn't need them before! :)

If you're curious, you can find the NULL sessions being used here:

- src/conn/conn_cache_pool.c:402:    `__wt_spin_lock(NULL, &cp->cache_pool_lock);`
- src/conn/conn_api.c:2717:    `__wt_spin_lock(NULL, &__wt_process.spinlock);`

quchenhao replied:
I get it now. Thanks.

pmacko86 commented on src/include/mutex_inline.h:
```diff
@@ -266,6 +292,40 @@ __wt_spin_unlock(WT_SESSION_IMPL *session, WT_SPINLOCK *t)
+ * WT_ASSERT_SPINLOCK --
+ *      Assert that the session owns the spinlock.
+ */
+#define WT_ASSERT_SPINLOCK(session, t) WT_ASSERT((session), __wt_spin_owns((session), (t)));
```
Rename to WT_ASSERT_SPINLOCK_OWNED or WT_ASSERT_HAVE_SPINLOCK? Or WT_REQUIRE_SPINLOCK?

pmacko86 commented:
I went through as many spinlocks as I could and added the relevant assertions.

I mostly just tried to "document" the current behavior: I added asserts to functions that are always called with the lock, even if the function does not necessarily need it, only skipping functions where I was reasonably confident that they probably don't need them. If you see any asserts that are not actually needed, please add comments with the explanations.

Conversely, if there are some functions that are missing asserts, please let me know—especially if it is in a major extern function. I only went one or two levels of call stack deep to add the asserts, so I'm sure there are a lot of places where I missed it.

I had a particularly difficult time with the schema lock, as there were many functions that were usually called with the lock, but I was not able to verify by reading the code whether they are always being called with the lock. For example, I wasn't able to do this for `__wt_schema_create`.

Another source of complications that would be good to address at some point are internal sessions—I don't have a good answer right now, except to just check whether those locks are being locked as opposed to whether the current session holds the lock. This may create additional problems with placing asserts further down the call stack. (Maybe we can find a way to do this better in a separate ticket? The scope of this ticket is already substantial.)

sueloverso commented on src/include/session.h:
```diff
@@ -53,6 +53,12 @@ typedef TAILQ_HEAD(__wt_cursor_list, __wt_cursor) WT_CURSOR_LIST;
+#define WT_SESSION_ID_INVALID 0xffffffff
+
+/* A fake session ID for when we need to refer to a session that is actually NULL. */
+#define WT_SESSION_ID_NULL 0xfffffffe
```
Given that you explicitly set the id to 64 bits, should these really be something like `include/log.h` where we have:
#define WT_LOG_SLOT_RESERVED 0x8000000000000000LL /* Reserved states */
#define WT_LOG_SLOT_MASK_OFF 0x3fffffffffffffffLL

I.e. explicitly use 64 bits in the id. And if you're creating a special value set them from the high bits in case we ever have a need to add a mask or bits, etc.

pmacko86 replied:
Our session IDs are actually 32-bit, even as the corresponding field in the `__wt_session_impl` struct is `uint32_t`, so this is meant to be a 32-bit constant in case it might become somewhere else.

I declared the session ID field in the mutex as `uint64_t` so that everything is properly 8-byte memory aligned, but now that I'm looking at the `__wt_spinlock` struct again, this is probably pointless as everything after that is `int16_t` or `int8_t`. Maybe I should just change the session ID field in `__wt_spinlock` to `uint32_t`.

sueloverso replied:
You could use uint32_t and if alignment is needed add a `uint8_t unused[4];` field.  I think if everything is `uint32_t` it would be best to have this also be.

pmacko86 replied:
Sounds good, I'll keep it in mind for next time. I don't think we need the alignment this time—now that I think about it, I'm not quite sure why I put it in there the first time.

sueloverso reviewed (approved):
LGTM with a few small comments. Let me know if you decide to separate out mechanism from usage. But overall this is nice.

ddanderson reviewed (approved):
LGTM!
Lots of great built in documentation as a result.

quchenhao commented on src/conn/conn_cache_pool.c:
```diff
@@ -207,6 +207,11 @@ __wt_cache_pool_config(WT_SESSION_IMPL *session, const char **cfg)
+    /*
+     * Cannot use __wt_spin_unlock_if_owned because we can reach this without the lock being
+     * initialized.
+     */
```
Can we fix the uninitialized lock?

pmacko86 replied:
We can't, because there are several places from which we can error out here before we even allocate the struct that contains the lock. But I just realized that we can do this more cleanly by using variable that tracks whether the lock exists (I'll call it `cp_lock_available`). Let me try that.

quchenhao commented on src/schema/schema_create.c:
```diff
@@ -1476,6 +1476,12 @@ __wt_schema_create(WT_SESSION_IMPL *session, const char *uri, const char *config
+     * This function seems to not always expect the schema lock to be acquired. For example, the
+     * lock is not acquired if we get here through __wt_tiered_open, but we still have to verify
+     * that.
+     */
```
There is no verification here?

pmacko86 replied:
I wasn't able to confirm that we always call this with the schema lock. I traced calls to this function about 4-5 levels deep, and one of those paths doesn't include acquiring the lock. Maybe it's acquired further up, but it's really hard to verify it as that function 4 or 5 levels up gets called from a large number of places, all without acquiring the lock. That being said, it's possible that this is a bug, but so far I was just documenting the current locking behavior.

pmacko86 replied:
That being said, I went ahead and added an assert here, even though I wasn't able to verify this through code reading. It passed both the csuite and the Python suite, so I hope it is indeed correct 👍 In the worst case, we'll get a test failure at some point, at which point we'll figure out what is the right thing to do. Needing schema lock at this point sounds reasonable.

pmacko86 commented on src/block/block_ckpt.c:
```diff
@@ -218,6 +218,8 @@ __wt_block_checkpoint_unload(WT_SESSION_IMPL *session, WT_BLOCK *block, bool che
 __wt_block_ckpt_destroy(WT_SESSION_IMPL *session, WT_BLOCK_CKPT *ci)
 {
+    /* Make sure to own the live lock when running this on the live checkpoint. */
+
```
This function is called for both live checkpoints with the lock, and for other checkpoints without the lock, and I didn't see a quick way to check within the function whether the checkpoint is live, which is why I left the comment instead of adding an assert.

FWIW, this function is also called from several places in a loop without the lock, and I'm not familiar enough with the code to tell whether any of those invocations could be with the live snapshot.

There is actually one place in `__wt_block_close` where we call this function without the lock: We first destroy the spinlock and then we call this to destroy the checkpoint. So strictly speaking, we are not always calling this function on the live checkpoint with the lock acquired.

How do you recommend that we should handle this?

1. Leave the comment as it is.
2. Remove the comment, if it's too confusing.
3. Find a way to check if this is the live checkpoint, so that we can add the assert (and update `__wt_block_close` accordingly). But for this option I would need a hint or two :)

sueloverso replied:
Then I would say something like:
We should hold the live lock here when running on the live checkpoint. But there is
no easy way to determine if the checkpoint is live so we cannot assert the locking here.

pmacko86 replied:
That's a good idea, let me do that.

pmacko86 commented on src/block/block_ext.c:
```diff
@@ -904,6 +915,8 @@ __wt_block_extlist_merge(WT_SESSION_IMPL *session, WT_BLOCK *block, WT_EXTLIST *
     u_int i;
 
+    /* Make sure to own the live lock spinlock if operating on the live snapshot. */
+
```
Same here. I haven't figured out a way yet to check if this is the live checkpoint.

quchenhao reviewed (approved):
lgtm
